overlord/state/warning.go

```diff
@@ -0,0 +1,282 @@
+	RepeatAfter string     `json:"repeat-after,omitempty"`
+}
+
+type Warning struct {
+	// the warning text itself. Only one of these in the system at a time.
+	message string
+	// the first time one of these messages was created
+	firstAdded time.Time
+	// the last time one of these was created
+	lastAdded time.Time
+	// the last time one of these was shown to the user
+	lastShown time.Time
+	// how much time since one of these was last added should we drop the message
+	expireAfter time.Duration
+	// how much time since one of these was last shown should we repeat it
+	repeatAfter time.Duration
+}
+
+func (w *Warning) String() string {
+	return w.message
+}
+
+func (w *Warning) MarshalJSON() ([]byte, error) {
+	jw := jsonWarning{
+		Message:     w.message,
+		FirstAdded:  w.firstAdded,
+		LastAdded:   w.lastAdded,
+		ExpireAfter: w.expireAfter.String(),
+		RepeatAfter: w.repeatAfter.String(),
+	}
+	if !w.lastShown.IsZero() {
+		jw.LastShown = &w.lastShown
+	}
+
+	return json.Marshal(jw)
+}
+
+func (w *Warning) UnmarshalJSON(data []byte) error {
+	var jw jsonWarning
+	err := json.Unmarshal(data, &jw)
+	if err != nil {
+		return err
+	}
+	w.message = jw.Message
+	w.firstAdded = jw.FirstAdded
+	w.lastAdded = jw.LastAdded
+	if jw.LastShown != nil {
+		w.lastShown = *jw.LastShown
+	}
+	if jw.ExpireAfter != "" {
+		w.expireAfter, err = time.ParseDuration(jw.ExpireAfter)
+		if err != nil {
+			return err
+		}
+	}
+	if jw.RepeatAfter != "" {
+		w.repeatAfter, err = time.ParseDuration(jw.RepeatAfter)
+		if err != nil {
+			return err
+		}
+	}
+
+	return w.validate()
+}
+
+func (w *Warning) validate() (e error) {
+	if w.message == "" {
+		return errNoMessage
+	}
+	if strings.TrimSpace(w.message) != w.message {
+		return errBadMessage
+	}
+	if w.firstAdded.IsZero() {
+		return errNoFirstAdded
+	}
+	if w.expireAfter == 0 {
+		return errNoExpireAfter
+	}
+	if w.repeatAfter == 0 {
+		return errNoRepeatAfter
+	}
+	return nil
+}
+
+func (w *Warning) ExpiredBefore(now time.Time) bool {
+	return w.lastAdded.Add(w.expireAfter).Before(now)
+}
+
+func (w *Warning) ShowAfter(t time.Time) bool {
+	return (w.lastShown.IsZero() || w.lastShown.Add(w.repeatAfter).Before(t)) && !w.firstAdded.After(t)
+}
+
+// flattenWarning loops over the warnings map, and returns all
+// non-expired warnings therein as a flat list, for serialising.
+// Call with the lock held.
+func (s *State) flattenWarnings() []*Warning {
+	now := time.Now()
+	flat := make([]*Warning, 0, len(s.warnings))
+	for _, w := range s.warnings {
+		if w.ExpiredBefore(now) {
+			continue
+		}
+		flat = append(flat, w)
+	}
+	return flat
+}
+
+// unflattenWarnings takes a flat list of warnings and replaces the
+// warning map with them, ignoring expired warnings in the process.
+// Call with the lock held.
+func (s *State) unflattenWarnings(flat []*Warning) {
+	now := time.Now()
+	s.warnings = make(map[string]*Warning, len(flat))
+	for _, w := range flat {
+		if w.ExpiredBefore(now) {
+			continue
```

Comment:
is this bit tested?

Reply:
no; I'd have to play with times a little too much (it'd have to not have expired during serialise, and then expire before unserialise; or I'd have to do it by hand? i guess i could do that).